fairscale/optim/layerwise_gradient_scaler.py

```diff
@@ -0,0 +1,278 @@
+        self._max_scaling_factor = max_scaling_factor
+        self._handles: List = []
+        self.layer_info: List = []
+
+        if self._apply_layerwise_scaling:
+            assert self._growth_factor > 1.0, "The growth factor must be > 1.0."
+            assert self._backoff_factor < 1.0, "The backoff factor must be < 1.0."
+            self.layer_info = self._build_layer_info()
+
+    def _build_layer_info(self) -> List:
+        """
+        Helper function to create a list of LayerInfo instances.
+        """
+        layer_info_list = list()
+
+        for name, layer in self._model.named_modules():
+            if name != "":
+                if name not in self._layer_scale_dict.keys():
+                    logging.debug("name = %s, layer = %s, scaling_factor = %s" % (name, layer, 1.0))
+                    layer_info_list.append(LayerInfo(name, layer, 1.0))
+                else:
+                    logging.debug(
+                        "name = %s, layer = %s, scaling_factor = %s" % (name, layer, self._layer_scale_dict[name])
+                    )
+                    layer_info_list.append(LayerInfo(name, layer, self._layer_scale_dict[name], True))
+        return layer_info_list
+
+    def scale(self) -> None:
+        """
+        For each layer calculates the scaling factor for preceding layers' grad inputs
+        and current layers' grad outputs. These values are used to register a full backward
+        hook. The handle returned from registering the backward hook is appended to a list
+        of handles. New hooks are created and registered at every step and a new list of
+        handles is created. The handles are flushed out in the unscale function.
+        """
+        if self._apply_layerwise_scaling:
+            for idx in range(len(self.layer_info)):
+                elt = self.layer_info[idx]
+                layer_name, layer_type = elt.layer_name, elt.layer_type
+
+                inputs_multiplier = 1.0
+                if idx > 0:
+                    inputs_multiplier = self.layer_info[idx - 1].scaling_factor
+
+                outputs_multiplier = 1.0 / elt.scaling_factor
+                helper = GradientHelper(layer_name, inputs_multiplier, outputs_multiplier)
+                layer_handle = layer_type.register_full_backward_hook(helper.scale_gradients)
+                self._handles.append(layer_handle)
+                logging.debug("name = %s \t scale = %s" % (layer_name, elt.scaling_factor))
+
+    def unscale(self) -> None:
+        """
+        For each layer, check if any of the layers' parameters contain an inf/nan.
+        If there are no inf/nan in the gradient, then gradient of that layer is
+        unscaled by the reciprocal of the scaling factor for that layer.
+        Finally, all handles recorded while registering the hooks are deleted.
+        """
+        if self._apply_layerwise_scaling:
+            for elt in self.layer_info:
```

Comment:
Can we use a filter function to only get the layers_info objects that don't have inf or nans? That would improve the readability here I think.

Reply:
done